### ISSUE TITLE
Read on the case of json/josnb, still do not understand how to write jsonb and other data mixed, can add a little more complex JSON example?

### ISSUE THREAD

**Issue by lchjczw** [jackc/pgx#156]
Read on the case of json/josnb, still do not understand how to write jsonb and other data mixed, can add a little more complex JSON example?

**Comment by jackc** [comment]
Not sure I understand exactly what you are trying to do. Have you seen this example: https://github.com/jackc/pgx/blob/master/example_json_test.go

**Comment by lchjczw** [comment]
This example I have seen, but some do not understand, how the general variables and jsonb mixed together

**Comment by jackc** [comment]
Can you give me an concrete example of what you are trying to do?

**Comment by lchjczw** [comment]
-- object: public.massage_binary | type: TABLE --
-- DROP TABLE IF EXISTS public.massage_binary;
CREATE TABLE public.massage_binary(
	mid int4 NOT NULL,
	did int4 NOT NULL,
	subid int2 NOT NULL,
	head jsonb NOT NULL,
	data bytea NOT NULL,
	CONSTRAINT pmidb PRIMARY KEY (mid)

)
TABLESPACE pg_default;
-- ddl-end --
COMMENT ON TABLE public.massage_binary IS '设备数据表';
-- ddl-end --
COMMENT ON COLUMN public.massage_binary.mid IS '设备id';
-- ddl-end --
COMMENT ON COLUMN public.massage_binary.did IS '设备id';
-- ddl-end --
COMMENT ON COLUMN public.massage_binary.subid IS '子id';
-- ddl-end --
COMMENT ON COLUMN public.massage_binary.head IS '数据';
-- ddl-end --
ALTER TABLE public.massage_binary OWNER TO postgres;
-- ddl-end --
------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------
I want to insert the data into this table。

	_mid int4 NOT NULL,_
	_did int4 NOT NULL,_
	_subid int2 NOT NULL,_
	_head jsonb NOT NULL,_
	_data bytea NOT NULL,_

Head is the JSON format, data is a binary format，I would like to know how to operate this mixed data？

**Comment by jackc** [comment]
`json` or `jsonb` will just be treated as another column, only the JSON will automatically be serialized and deserialized into the `struct` or `map`. So you would insert the row the same as any other multiple column row.

It'd be something like the following:

```go
type head struct {
	Name string `json:"name"`
	Age  int    `json:"age"`
}

mid := int32(1)
did := int32(2)
subid := int16(3)

h := head{
	Name: "John",
	Age:  42,
}

data := []byte{1, 2, 3, 4, 5}

_, err := conn.Exec("insert into massage_binary values($1, $2, $3, $4, $5)",
	mid, did, subid, head, data)
if err != nil {
	return err
}
```

**Comment by lchjczw** [comment]
think you！

**Comment by lchjczw** [comment]
_, err := conn.Exec("insert into massage_binary values($1, $2, $3, $4, $5)",
    mid, did, subid, head, data)
----------------------------------------------------------------------
About head variables, can I pass in the JSON string? Because JSON this place will change, if the parameter is the structure, then the flexibility is not high enough

**Comment by josephglanville** [comment]
@lchjczw A whole bunch of types can automatically be encoded to `json` or `jsonb` including `struct`, `*struct`, `interface`, `*interface`. If you want do to the encoding/decoding yourself you can supply `string`, `*string` or `json.RawMessage` or `*json.RawMessage`.

**Comment by lchjczw** [comment]
I mean that when I store data, do not need to be converted to JSON through the structure or other types, directly through the JSON string

**Comment by josephglanville** [comment]
Then just supply a `string` to `Exec()`/`Query()` etc, pgx won't attempt any serialisation in that case.

**Comment by agentabstract** [comment]
I seriously haven't been able to get this to work at all.  Is there a certain version of postgresql required?  My field in the table is jsonb. If I Marshall the value into bytes, I get version 123 not supported.  If I pass the exec command a map or a struct variable I get a syntax error from Postgres. I'm away from my IDE. I was trying to research the problem on my phone,  but I can give exact error messages later.

**Comment by agentabstract** [comment]
Btw- so far I love the library. This is the first problem I've had, and I'm so close to being done with a data converter for my company. It's just a big part of what we're doing is counting on jsonb fields.  I'm totally stuck.

**Comment by jackc** [comment]
That is very odd. Doesn't make sense that which package it is called from would have any effect. I wonder if there is some connection state that is different between the connections.

That error looks like its underlying type would be a pgx.PgError. Can you try a type assertion on the error and convert to pgx.PgError? That error may have a lot more details. (https://godoc.org/github.com/jackc/pgx#PgError)

**Comment by agentabstract** [comment]
Inserting this into my transaction loop: 
`				if err != nil{
					var pgxerror pgx.PgError
					pgxerror = err.(pgx.PgError)
					fmt.Println(pgxerror.Error())
				}`
Resulted in: 
`ERROR: invalid input syntax for type json (SQLSTATE 22P02)
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
ERROR: current transaction is aborted, commands ignored until end of transaction block (SQLSTATE 25P02)
panic: commit unexpectedly resulted in rollback`

**Comment by agentabstract** [comment]
I figured it out. 

Originally all of the fields of the struct that were jsonb were type string.  All of my initial tests of the importer were working perfectly with this setup. All of those fields were not being assigned, but were in all of my update/insert statements.  I started to assign my first jsonb field, and in doing so I changed the type to map[string]string,  but left all of the others as string.  None of my tests worked.  I came to you looking for help.  In troubleshooting I commented out the lines where I assigned values to the single field. It was now unassigned, and I got the same errors. I decided to try reverting even further, and defined the field not as map, but again to a string.  Everything should work as it had during my initial tests, but I received the same error.  I said many things a person shouldn't say aloud, and threw a pen at the wall in front of me.  There in my moment of anger, I had an epiphany. I redefined all of the jsonb fields in the struct as type map[string]string.  Now I'm not getting any errors.
